compiler/modules/bir/verify.bal

```diff
@@ -65,13 +65,223 @@ class VerifyContext {
+    BasicBlock[] blocks = code.blocks;
+    int numBlocks = blocks.length();
+    if numBlocks == 0 {
+        return cx.invalidErr("no basic blocks for function", funcPos);
+    }
+    // Checks are done in two DFS rounds. During first round, forward in degree is calculated for each block.
+    int[] fwdInDegree = [];
+    fwdInDegree[numBlocks - 1] = 0;
+    boolean[] preds = [];
+    preds[numBlocks - 1] = false;
+    check verifyStructure(cx, blocks, 0, fwdInDegree, preds, funcPos);
+    check verifyReachable(cx, fwdInDegree, funcPos);
+    // In second round, previously calculated forward in degree is used as an input.
+    boolean[] paramRegs = check verifyParamRegs(cx, code.registers);
+    RegFlow[][] blocksFlows = [];
+    blocksFlows[numBlocks - 1] = [];
+    check verifyRegFlow(cx, blocks, 0, fwdInDegree, blocksFlows, (), paramRegs, funcPos);
+
+    foreach BasicBlock b in blocks {
         check verifyBasicBlock(cx, b);
     }
 }
 
+// Verify forward edgers form a DAG
+// Verify backward branch points to a predecessor XXX: need to also verify it's a dominator
+// Calculates forward in-degree for each block
+function verifyStructure(VerifyContext vc, BasicBlock[] blocks, Label current, int[] fwdInDegree, boolean[] preds, Position predPos) returns Error? {
+    BasicBlock block = blocks[current];
+    preds[current] = true;
+    Insn term = check terminator(vc, block.insns, predPos);
+    Position termPos = term.pos;
+    Label? onPanic = block.onPanic;
+    if onPanic != () {
+        check verifyChildStructure(vc, blocks, onPanic, fwdInDegree, preds, termPos);
+    }
+    if term is BranchInsn {
+        check verifyChildStructure(vc, blocks, term.dest, fwdInDegree, preds, termPos, term.backward);
+    }
+    else if term is TypeBranchInsn|CondBranchInsn {
+        check verifyChildStructure(vc, blocks, term.ifTrue, fwdInDegree, preds, termPos);
+        check verifyChildStructure(vc, blocks, term.ifFalse, fwdInDegree, preds, termPos);
+    }
+    preds[current] = false;
+    fwdInDegree[current] = 1; // fwdInDegree[current] was 0 until now since above recursions can't cycle back to self.
+}
+
+function verifyChildStructure(VerifyContext vc, BasicBlock[] blocks, Label child, int[] fwdInDegree, boolean[] preds, Position pos, boolean backward = false) returns Error? {
+    boolean cycle = preds[child];
+    if backward != cycle {
+        return vc.invalidErr(backward ? "backward branch to non-predecessor" : "forward branch form a cycle", pos);
+    }
+    if backward {
+        return;
+    }
+    int fid = fwdInDegree[child];
+    if fid == 0 { // unvisited
+        check verifyStructure(vc, blocks, child, fwdInDegree, preds, pos);
+    }
+    else {
+        fwdInDegree[child] = fid + 1;
+    }
+}
+
+function verifyReachable(VerifyContext vc, int[] fwdInDegree, Position pos) returns Error? {
+    foreach int fid in fwdInDegree {
+        if fid == 0 {
+            return vc.invalidErr("block is not reachable", pos);
```

Comment:
Don't we want the position of the unreachable block?

Reply:
Yes, most of the time. I was considering the case where it may be empty. Added a check.